cpp/unreal_projects/SpearSim/Python/validate_spear_scene.py

```diff
@@ -0,0 +1,579 @@
+        # should be assigned to whatever subdirectory is most convenient, i.e., requires the least
+        # number of per-component overrides.
+        #
+        # We choose to encode semantic annotations via this directory structure, because it
+        # makes it especially easy to browse each scene in the Unreal Editor. For example,
+        # using this directory structure, a user can easily select all objects in a particular
+        # semantic category and make them invisible, thereby making it easier to browse the
+        # rest of the scene.
+        #
+
+        semantic_category = path[1].split("_")
+
+        if len(semantic_category) < 2:
+            spear.log("WARNING: Unexpected path: ", get_debug_string_actor(actor))
+            return
+
+        if len(semantic_category[0]) != 4 or not semantic_category[0].isdigit():
+            spear.log("WARNING: Unexpected path: ", get_debug_string_actor(actor))
+            return
+
+        if int(semantic_category[0]) == 0:
+            spear.log("WARNING: Unexpected path: ", get_debug_string_actor(actor))
+            return
+
+        actor_label = actor.get_actor_label().split("_")
+        if actor_label[:-1] != semantic_category[1:]:
+            spear.log("WARNING: Unexpected path: ", get_debug_string_actor(actor))
+            return
+
+        #
+        # The type of each actor in my_scene_0000/Meshes should be StaticMeshActor if it is not
+        # articulated, and Actor if it is articulated. It is necessary to use the Actor type for
+        # articulated actors so they can be correctly physically simulated. Although it is not
+        # strictly necessary to use the StaticMeshActor type for non-articulated actors,
+        # StaticMeshActors and Actors have different icons in the Outliner pane, so we use
+        # different types to visually differentiate them.
+        #
+
+        if type(actor) != unreal.Actor and type(actor) != unreal.StaticMeshActor:
+            spear.log("WARNING: Unexpected type: ", get_debug_string_actor(actor), " (", get_debug_string_type(type(actor)), ")")
+            return
+
+        #
+        # The pivot location of each StaticMeshActor should be set according to the
+        # following rules. The xy-coordinates of the actor's pivot should equal the
+        # xy-coordinates of the actor's axis-aligned bounding box center, and the
+        # z-coordinate of the pivot should equal the minimum z-coordinate of its
+        # axis-aligned bounding box. Having a consistent convention here makes it easier
+        # to programmatically spawn objects, and this is the convention for various props
+        # that ship with the Unreal Engine (e.g., the props in our "debug_0000" scene).
+        #
+        # All the rules for StaticMeshActors described above also apply to Actors.
+        #
+
+        # ignore components from bounds computation if they don't have a mesh assigned, or if their assigned mesh is SM_Dummy
```

Comment:
I just realized we need to handle the case when there is rotation and scaling being applied to the actor. The updated implementation below reduces to my original implementation in the case of identity rotation and scaling, but it handles non-uniform rotation and scaling in a sensible way.

```
        # ignore components from bounds computation if they don't have a mesh assigned, or if their assigned mesh is SM_Dummy
        ignore_components = [
            c for c in actor.get_components_by_class(unreal.StaticMeshComponent) if
            c.get_editor_property("static_mesh") is None or
            c.get_editor_property("static_mesh").get_path_name() == "/Game/Kujiale/Meshes/SM_Dummy.SM_Dummy"]

        # cache the actor's transform
        transform_world_from_obj = actor.get_actor_transform()

        # for each ignored component, cache "use_default_collision" and "collision_enabled" properties
        ignore_use_default_collision = {
            c.get_name(): c.get_editor_property("use_default_collision") for c in ignore_components}
        ignore_collision_enabled = {
            c.get_name(): c.get_editor_property("body_instance").get_editor_property("collision_enabled") for c in ignore_components}

        # set the actor's transform to be the identity transform because we want to compute bounds in object space
        actor.set_actor_transform(unreal.Transform(), sweep=False, teleport=True)

        # disable collision on each ignored component
        for c in ignore_components:
            c.set_editor_property("use_default_collision", False)
            c.get_editor_property("body_instance").set_editor_property("collision_enabled", unreal.CollisionEnabled.NO_COLLISION)

        # compute bounds
        bounds_origin_obj, bounds_half_extent_obj = actor.get_actor_bounds(only_colliding_components=True, include_from_child_actors=False)

        # for each ignored component, restore "use_default_collision" and "collision_enabled" properties
        for c in ignore_components:
            c.set_editor_property("use_default_collision", ignore_use_default_collision[c.get_name()])
            c.get_editor_property("body_instance").set_editor_property("collision_enabled", ignore_collision_enabled[c.get_name()])

        # restore the actor transform
        actor.set_actor_transform(transform_world_from_obj, sweep=False, teleport=True)

        # compute the desired pivot location in world space
        pivot_desired_obj = unreal.Vector(bounds_origin_obj.x, bounds_origin_obj.y, bounds_origin_obj.z - bounds_half_extent_obj.z)
        pivot_desired_world = transform_world_from_obj.transform_location(pivot_desired_obj)

        # check pivot location against computed bounds
        eps = 0.000001
        pivot_desired_world = vector_to_array(pivot_desired_world)
        pivot_actual_world = vector_to_array(actor.get_actor_location())
        if np.linalg.norm(pivot_desired_world - pivot_actual_world) > eps:
            np.set_printoptions(suppress=True)
            spear.log(
                "WARNING: Unexpected pivot: ", get_debug_string_actor(actor),
                " (desired=", pivot_desired_world, ", actual=", pivot_actual_world, ")")
            return
```

Reply:
In this screenshot, notice how the white sphere is positioned at a sensible desired pivot location (`pivot_desired_world`) at the base of the table, even though the `StaticMeshComponent` that references the table mesh is very off-center with respect to its current pivot location (`pivot_actual_world`), and has non-uniform rotation and scaling applied to the actor.

![Screenshot (79)](https://github.com/isl-org/spear/assets/2341965/296b4837-da6a-426f-9167-42a6608a8328)